src/plot_api/plot_api.js

```diff
@@ -1580,6 +1580,12 @@ function _restyle(gd, aobj, _traces) {
                 var moduleAttrs = (contFull._module || {}).attributes || {};
                 var valObject = Lib.nestedProperty(moduleAttrs, ai).get() || {};
 
+                // if restyling entire attribute container, assume worse case
+                if(!valObject.valType) {
```

Comment:
A little bit hacky as this relies on the fact that attribute containers (e.g. `marker`) don't have a `valType` field set in the attribute declarations. 

Morevoer, we could optimize this further by digging into the attribute container for `arrayOk` keys, but I thought this wasn't necessary. For best-performing `restyle` calls, users should target specific attributes e.g. `restyle(gd, 'marker.color', [['red', 'blue', 'green']])`.